File: 2074-reverse-nodes-in-even-length-groups/2074-reverse-nodes-in-even-length-groups.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    int k=1;
    public ListNode reverse(ListNode start,ListNode end){
        ListNode front=start,ptr=start,back=null;
        while(front!=end){
            front=front.next;
            ptr.next=back;
            back=ptr;
            ptr=front;
        }
        return back;
    }
    
    public ListNode reverseEvenLengthGroups(ListNode head) {
        ListNode node=head,prev=null;
        for(int i=0;i<k;i++){
            if(node==null){
                if(i%2==0){
                    return reverse(head,node);
                }
                return head;
            }
            prev=node;
            node=node.next;
        }
    
        if(k%2!=0){
            k++;
            prev.next=reverseEvenLengthGroups(node);
            return head;
        }
        k++;
        ListNode nNode=reverse(head,node);
        head.next=reverseEvenLengthGroups(node);
        return nNode;
    }
}

File: 2074-reverse-nodes-in-even-length-groups/2074-reverse-nodes-in-even-length-groups.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode*reverse(ListNode*start,ListNode*end){
        ListNode*st=start,*last=NULL,*ptr=start;
        while(st!=end){
            st=st->next;
            ptr->next=last;
            last=ptr;
            ptr=st;
        }
        return last;
    }
    ListNode* reverseEvenLengthGroups(ListNode* head,int k=1) {
        
        ListNode*node=head,*prev;
        for(int i=0;i<k;i++){
            if(!node){
                if(i%2==0){
                    return reverse(head,NULL);
                }
                return head;
            }
            prev=node;
            node=node->next;
        }
        if(k%2!=0){
         prev->next= reverseEvenLengthGroups(node,k+1);
            return head;
        }
        ListNode*nNode=reverse(head,node);
        head->next=reverseEvenLengthGroups(node,k+1);
        return nNode;
        
    }
};